overlord/devicestate/users.go

```diff
@@ -0,0 +1,372 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package devicestate
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os/user"
+	"path/filepath"
+	"time"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/overlord/assertstate"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/strutil"
+)
+
+type CreatedUser struct {
+	Username string
+	SSHKeys  []string
+}
+
+var (
+	osutilAddUser = osutil.AddUser
+	osutilDelUser = osutil.DelUser
+	userLookup    = user.Lookup
+)
+
+type UserError struct {
+	Err      error
+	Internal bool
+}
+
+func (e *UserError) Error() string {
+	return e.Err.Error()
+}
+
+func (e UserError) IsInternal() bool {
+	return e.Internal
+}
+
+// RemoveUser removes linux user account of passed username.
```

Comment:
the doc comment of these functions need to explain that they take the lock themselves, or switch to assume the lock is taken and unlock as needed. Because taking the lock for top level functions in manager that take state is the unusual pattern. @Meulengracht given your use case we maybe should do the second thing?

Reply:
Yea, just like you said. I changed how the locking is done, and now the functions assumes the lock like other state API's in the manager. This will also fit better given the other use case